Libraries/oneMKL/computed_tomography/computed_tomography.cpp

```diff
@@ -407,181 +254,675 @@ void bmp_write_templ(std::string fname, int h, int w, int ldw, T *data)
+    const int reconstruction_ldw = reconstruction.ldw();
+    const int original_ldw = original.ldw();
+    auto compute_mean_error_in_pixel = [=](int i, int j) {
+        // Notes:
+        // - pixel of index (i,j) in the reconstructed image covers the area of
+        //   points (y, x) such that
+        //              |y - (i - q/2)*pixel_length| <= 0.5*pixel_length
+        //              |x - (j - q/2)*pixel_length| <= 0.5*pixel_length
+        // - pixel of index (io,jo) in the original image covers the area of
+        //   points (y, x) such that
+        //          io*in_pix_len <= y + 0.5*hh <= (io+1)*in_pix_len
+        //          jo*in_pix_len <= x + 0.5*ww <= (jo+1)*in_pix_len
+        // in a common orthonormal reference frame centered on either image
+        const double y_min = (i - q/2 - 0.5)*pixel_length;
+        const double x_min = (j - q/2 - 0.5)*pixel_length;
+        const double y_max = (i - q/2 + 0.5)*pixel_length;
+        const double x_max = (j - q/2 + 0.5)*pixel_length;
+        if (y_min > 0.5*hh || y_max < -0.5*hh ||
+            x_min > 0.5*ww || x_max < -0.5*ww) {
+            // out of scope of relevance, no intersection with the original
+            // image
+            return 0.0;
+        }
+        // find corresponding range of pixels in original image
+        const auto io_jo_min = get_original_index(y_min, x_min);
+        const auto io_jo_max = get_original_index(y_max, x_max);
+        const double got = reconstruction_data[i*reconstruction_ldw + j];
+        double mean_local_error = 0.0;
+        // parse original image's pixels (io, jo) having a non-empty
+        // intersection with the reconstructed image's pixel (i, j)
+        for (int io = io_jo_min.first; io <= io_jo_max.first; io++) {
+            const double yo_min = -0.5*hh + io*in_pix_len;
+            const double yo_max = -0.5*hh + (io + 1)*in_pix_len;
+            double overlap_y =
+                sycl::min(y_max, yo_max) - sycl::max(y_min, yo_min);
+            overlap_y = sycl::max(0.0, sycl::min(overlap_y, max_overlap));
+            for (int jo = io_jo_min.second; jo <= io_jo_max.second; jo++) {
+                const double xo_min = -0.5*ww + jo*in_pix_len;
+                const double xo_max = -0.5*ww + (jo + 1)*in_pix_len;
+                double overlap_x =
+                    sycl::min(x_max, xo_max) - sycl::max(x_min, xo_min);
+                overlap_x = sycl::max(0.0, sycl::min(overlap_x, max_overlap));
+                const double exp = original_data[io*original_ldw + jo];
+                mean_local_error +=
+                    sycl::fabs(exp - got)*overlap_y*overlap_x;
+            }
+        }
+        mean_local_error /= (pixel_length*pixel_length);
+        return mean_local_error;
+    };
+    // compute the mean local errors and the mean global error
+    double mean_error = 0.0;
+    double* L1_error = (double *)sycl::malloc_device(sizeof(double), errors.q);
```

Comment:
```suggestion
    double* L1_error = sycl::malloc_device<double>(1, errors.q);
```
I find the templated versions a little more readable personally.

Reply:
Changed in 60611f839df20fa66ed8f3ca3febec3912c19eeb.